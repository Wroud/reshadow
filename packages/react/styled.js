import styled, {css, jsx, getDisplayName} from './';
import tags from '@reshadow/utils/html-tags';

const wrap = (element, arr) => {
    arr[0] = `${element} {` + arr[0];
    arr[arr.length - 1] = arr[arr.length - 1] + '}';
    return arr;
};

const createStyled = tag => (strs, ...values) => {
    const element = typeof tag === 'function' ? getDisplayName(tag) : tag;
    const wrapper = wrap(element, [...strs]);
    const newStrs = Object.create(wrapper);
    newStrs.raw = newStrs;

    return props => styled(
        newStrs,
        ...values.map(x => (typeof x === 'function' ? x(props) : x)),
    )(jsx(tag, props));
};

const fn = Base => createStyled(Base);

tags.forEach(tag => {
    fn[tag] = createStyled(tag);
});

export {css};

export default fn;
