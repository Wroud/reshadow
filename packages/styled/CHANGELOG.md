# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.0.1-alpha.47](https://github.com/lttb/reshadow/compare/v0.0.1-alpha.46...v0.0.1-alpha.47) (2019-05-30)

**Note:** Version bump only for package @reshadow/styled





## [0.0.1-alpha.46](https://github.com/lttb/reshadow/compare/v0.0.1-alpha.45...v0.0.1-alpha.46) (2019-05-30)


### Bug Fixes

* **styled:** use global selectors ([3ed8e46](https://github.com/lttb/reshadow/commit/3ed8e46))





## [0.0.1-alpha.45](https://github.com/lttb/reshadow/compare/v0.0.1-alpha.44...v0.0.1-alpha.45) (2019-05-30)


### Features

* **runtime:** add keyframes ([5f19687](https://github.com/lttb/reshadow/commit/5f19687))





## [0.0.1-alpha.44](https://github.com/lttb/reshadow/compare/v0.0.1-alpha.43...v0.0.1-alpha.44) (2019-05-30)


### Features

* **styled:** improve compatibility with SC API ([78304ca](https://github.com/lttb/reshadow/commit/78304ca))
