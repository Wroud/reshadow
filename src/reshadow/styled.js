const isFalsy = val =>
    val === undefined || val === null || val === false || val === '';

const __id__ = Symbol('__id__');
const __store__ = Symbol('__store__');
const __prev__ = Symbol('__prev__');
const __use__ = Symbol('__use__');

let index = 0;

const PREFIX = 'use--';

const use = obj => {
    const result = {};
    result[__use__] = obj;
    return result;
};

const create = args => {
    const len = args.length;
    let newStyles = {};
    let id = '';

    for (let i = 0; i < len; i++) {
        let style = args[i];

        if (!style) continue;

        if (!(__id__ in style)) {
            style[__id__] = ++index;
            style[__store__] = {
                ['_' + style[__id__]]: style,
            };
        }

        id += '_' + style[__id__];

        if (style[__store__][id]) {
            newStyles = style[__store__][id];
            continue;
        }

        newStyles = Object.create(newStyles);

        for (let key in style) {
            if (newStyles[key]) {
                newStyles[key] += ' ';
            } else {
                newStyles[key] = '';
            }
            newStyles[key] += style[key];
        }

        style[__store__][id] = newStyles;
        newStyles[__id__] = index;
    }

    return newStyles;
};

let styles = {};

const styled = elem => {
    if (styles[__prev__]) {
        styles = styles[__prev__];
        styles[__prev__] = null;
    }

    styled.styles = styles;

    return elem;
};

const css = (code, hash) => {
    let container = document.getElementById('reshadow');
    if (!container) {
        container = document.createElement('object');
        container.id = 'reshadow';
        document.head.appendChild(container);
    }
    const id = `reshadow-${hash}`;
    let css = document.getElementById(id);
    if (!css) {
        css = document.createElement('style');
        css.id = id;
        css.type = 'text/css';
        container.appendChild(css);
    }

    css.innerHTML = code;
};

const set = args => {
    const newStyles = create(args);

    newStyles[__prev__] = styles;

    styles = newStyles;
    styled.styles = styles;
};

/**
 * This prop is needed for the interop between different component frameworks
 * TODO: think about better solution
 */
styled.classProp = 'className';

function map(element) {
    let nextProps = {};
    let cn = styles[`__${element}`] || '';
    let style = null;

    const len = arguments.length;

    for (let i = len - 1; i > 0; i--) {
        let currProps = arguments[i];

        let isUse = false;

        if (currProps && __use__ in currProps) {
            isUse = true;
            currProps = currProps[__use__];
        }

        if (!currProps) continue;

        for (let key in currProps) {
            if (key === '__style__') {
                style = currProps.__style__;
                continue;
            }

            const value = currProps[key];

            if (isUse) {
                key = PREFIX + key;
            } else if (key in nextProps) {
                continue;
            } else {
                nextProps[key] = value;
            }

            if (isFalsy(value)) continue;

            let className = styles[`_${key}`];
            if (className) {
                cn += (cn ? ' ' : '') + className;
            }

            if (typeof value !== 'boolean') {
                className = styles[`_${key}_${value}`];
                if (className) {
                    cn += (cn ? ' ' : '') + className;
                }
            }
        }
    }

    if (nextProps[styled.classProp]) {
        cn += (cn ? ' ' : '') + nextProps[styled.classProp];
    }

    if (cn) nextProps[styled.classProp] = cn;

    if (style) {
        nextProps.style =
            typeof style === 'string'
                ? style + (nextProps.style || '')
                : Object.assign(style, nextProps.style || {});
    }

    return nextProps;
}

export {use, css, create, set, map, css as __css__};

export default Object.assign(styled, {
    styles,
    use,
    css,
    create,
    set,
    map,
});
