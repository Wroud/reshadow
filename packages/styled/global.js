import React from 'react';
import {
    createStyled as createReshadowStyled,
    createCSS,
} from '@reshadow/runtime';
import coreStyled, {KEYS} from '@reshadow/core';
import stylis from 'stylis';
import styled from '.';

const globalCSS = createCSS({
    parse: code => {
        stylis.use(null);
        return {css: stylis('', code), tokens: {}};
    },
});

class GlobalStyle extends React.Component {
    render() {
        // eslint-disable-next-line
        const {style} = this.props;
        if (style) {
            for (let name in style) {
                document.documentElement.style.setProperty(name, style[name]);
            }
        }

        return null;
    }
}

const globalReshadowStyled = createReshadowStyled((as, props) => {
    let style = coreStyled[KEYS.__style__];

    props.style = style;

    return React.createElement(GlobalStyle, props);
}, globalCSS);

export const createGlobalStyle = (strs, ...values) =>
    styled(GlobalStyle).create({
        wrap: (element, strs) => strs,
        styled: globalReshadowStyled,
    })(strs, ...values);
