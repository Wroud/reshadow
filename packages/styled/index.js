import React from 'react';
import {getDisplayName} from '@reshadow/react';
import {
    keyframes,
    createStyled as createReshadowStyled,
    wrap,
    createCSS,
} from '@reshadow/runtime';
import coreStyled, {KEYS, map} from '@reshadow/core';
import tags from '@reshadow/utils/html-tags';
import {ThemeContext} from 'theming';
import isReactProp from 'is-react-prop';

const ThemeConsumer = ThemeContext.Consumer;

const styledProps = ['as', 'theme'];
const blacklist = new Set([...styledProps, 'color']);

const filterProps = props => {
    const nextProps = {};
    const filtered = Object.create(styledProps);
    Object.keys(props).forEach(prop => {
        nextProps[prop] = props[prop];
        if (blacklist.has(prop) || !isReactProp(prop)) {
            filtered.push(prop);
        }
    });
    return {props: nextProps, filtered};
};

const __css__ = createCSS({
    elements: false,
    classes: false,
});

const re = /\w+:[\s\r\n]*$/;

function css() {
    const str = arguments[0];
    const functions = {};
    const args = [str];

    for (let i = 1, len = arguments.length; i < len; i++) {
        const value = arguments[i];
        args[i] = value;
        if (typeof value === 'function') {
            functions[i] = value;
        } else if (typeof value === 'string' && !re.test(str[i - 1])) {
            args[i] = __css__([value]);
        }
    }

    const keys = Object.keys(functions);

    if (keys.length === 0) {
        return __css__.apply(null, args);
    }

    return data => {
        const nextArgs = Object.create(args);
        for (let index in functions) {
            let value = functions[index](data);
            if (typeof value === 'string' && !re.test(str[index - 1])) {
                value = __css__([value]);
            }

            while (typeof value === 'function') value = value(data);

            nextArgs[index] = value;
        }
        return __css__.apply(null, nextArgs);
    };
}

const reshadowStyled = createReshadowStyled((element, as, props) => {
    let style = coreStyled[KEYS.__style__];

    if (style) {
        props.style = Object.assign({}, style, props.style);
    }

    const result = React.createElement(as, map(element, props));

    if (style && result.props.style === props.style) {
        return result;
    }

    return React.cloneElement(result, {
        style: Object.assign({}, style, result.props.style),
    });
}, css);

const createStyled = tag => {
    const create = (options = {}) => (strs, ...values) => {
        const element = typeof tag === 'function' ? getDisplayName(tag) : tag;
        const elementClassName = '.__root__';
        const wrapper = (options.wrap || wrap)(elementClassName, [...strs]);

        class StyledComponent extends React.Component {
            constructor() {
                super();
                this.innerRender = this.innerRender.bind(this);
            }

            mapValues(props) {
                const result = [];
                for (let i = 0; i < values.length; i++) {
                    let value = values[i];

                    while (typeof value === 'function') value = value(props);

                    result.push(value);
                }
                return result;
            }

            innerRender(contextTheme) {
                const {props, filtered} = filterProps(this.props);

                if (options.attrs) {
                    for (let attr in options.attrs) {
                        const value = options.attrs[attr];
                        props[attr] =
                            typeof value === 'function'
                                ? value(this.props)
                                : value;
                    }
                }
                props.theme = props.theme || contextTheme;
                const as = props.as || tag;
                const localValues = this.mapValues(props);

                let localWrapper = wrapper;
                // eslint-disable-next-line
                let localMixin = this.props.css;

                if (localMixin) {
                    localWrapper = wrap(elementClassName, [...strs, '']);

                    while (typeof localMixin === 'function') {
                        localMixin = localMixin(props);
                    }

                    localValues.push(
                        typeof localMixin === 'string'
                            ? css([localMixin])
                            : localMixin,
                    );
                }

                const args = [localWrapper];

                args.push.apply(args, localValues);
                for (let i = 0; i < filtered.length; i++) {
                    delete props[filtered[i]];
                }
                return (options.styled || reshadowStyled).apply(null, args)(
                    'root__',
                    as,
                    props,
                );
            }

            render() {
                return React.createElement(
                    ThemeConsumer,
                    null,
                    this.innerRender,
                );
            }
        }

        StyledComponent.displayName = `styled.${element}`;

        StyledComponent.withComponent = name =>
            createStyled(name)(strs, ...values);

        StyledComponent.styledComponentId = 'id';

        return StyledComponent;
    };

    const result = create();

    result.attrs = attrs => create({attrs});
    result.create = create;

    return result;
};

const styled = Base => createStyled(Base);

tags.forEach(tag => {
    styled[tag] = createStyled(tag);
});

export * from 'theming';
export {createGlobalStyle} from './global';

export function isStyledComponent(target) {
    return target && typeof target.styledComponentId === 'string';
}

export {css, keyframes, ThemeConsumer};

export default styled;
