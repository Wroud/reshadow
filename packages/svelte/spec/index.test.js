const {stripIndent} = require('common-tags');

const reshadow = require('../preprocess');

const createTransform = (options = {}) => {
    const processor = reshadow(options);

    return content => processor.markup({content: stripIndent(content)});
};

const transform = createTransform();

describe('svelte preprocess', () => {
    it('should transform the code', () => {
        const code = transform`
            <script>
                import styled from 'reshadow'

                export let disabled = false
                export let size = 'm'

                styled\`
                    h1 {
                        color: \${color\};
                    }
                \`
            </script>

            <svelte:body on:click={() => console.log('body click')} />

            <h1>hello world</h1>

            <button :variant="action" :{size} {disabled}>click me</button>
        `;

        expect(code).toMatchSnapshot();
    });
});
