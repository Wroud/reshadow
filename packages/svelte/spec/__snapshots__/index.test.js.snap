// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`svelte preprocess should transform the code 1`] = `
Object {
  "code": "<script>import styled, { __extract__, set, create, css, use, map, __css__ } from \\"@reshadow/core\\";
export let name = 'name';
export let disabled = false;
export let size = 'm';
export let variant = 'action';
const styled_c8 = create([(__css__(\`.___h1_rzsno_1 {
    color: var(--c8_0);
}\`
/*__css_end__*/
, \\"1692710358\\"), {
  \\"__h1\\": \`___h1_rzsno_1\`
})]);
$: __styles__ = styled((set([styled_c8], \`--c8_0:\${color};\`), __extract__()));
;require('@reshadow/svelte').__init__(require('svelte'), () => __styles__, () => {
                styled = styled;
                map = map;
            });
            </script><style></style>

  <svelte:body class={styled.styles[\\"__svelte:body\\"]} style={__styles__.$$style} on:click={() => console.log('body click')} />

  <h1 class={styled.styles[\\"__h1\\"]} style={__styles__.$$style}>hello {name}</h1>

  <button {...map(\\"button\\", {
    disabled: disabled,
    $$style: __styles__.$$style
  }, use({
    state: \\"pending\\",
    variant: variant,
    size: size
  }))}>click me</button>",
}
`;
